### PR TITLE
Refactor stems/beams

### PR DESCRIPTION
**Pull request by Silverwolf90**
Still  a WIP, tab is broken.

Important things to note:

* The logic `StaveNote.getStemExtents()` has been removed and now delegates to `Stem.getExtents()`
* Went through `Beam` and did a lot of renaming, most variable names were really vague and I think it's a lot easier to understand now.
* A lot of logic in `beam.applyStemExtensions` was just plain wrong mainly due to incorrect assumptions around stem coordinates because it's a [very confusing part of the codebase](https://github.com/0xfe/vexflow/wiki/Development-Gotchas) -- that method has been greatly simplified
* `Formatter.FormatAndDraw` now passes `options.stave` to `formatter.formatToStave` which is required to trigger the `postFormat`  lifecycle (this also explains why the stem extents were wrong in the articulation PR)

### CODE REVIEW

**Comment by gristow** [comment]
This is great, so much easier to understand what's happening now!

**Comment by aaronmars** [comment]
@Silverwolf90 , do you think this could help with my tuplet headaches in https://github.com/0xfe/vexflow/issues/372 ?

**Comment by Silverwolf90** [comment]
Branches merge together without conflicts. I took the group of notes with the misaligned bracket:

Here's the test you added:

image

Here's that group of notes with previously mis-aligned tuplet:

image

**Comment by aaronmars** [comment]
That is wonderful news.  I'll stop trying to fix it with the current code and wait until your changes here are merged.

**Comment by 0xfe** [comment]
Thanks for the PR, Cyril. I'm traveling right now, and will review as soon as I get a chance.

**Comment by 0xfe** [comment]
Most of the tests have reasonable diffs in the regression test, but this one was really weird:

StaveNote.Center_Aligned_Note_with_Articulation (blessed, current, diff):

stavenote center_aligned_note_with_articulation_blessed
stavenote center_aligned_note_with_articulation_current
stavenote center_aligned_note_with_articulation

**Comment by Silverwolf90** [comment]
Looks like it fixed a bug :)

The test's title is Center_Aligned_Note_with_Articulation, and the blessed image doesn't have an articulation, while the current one has a fermata!

**Comment by Silverwolf90** [comment]
But Tab stems are stil broken. I started digger further and it seems that the `StaveNote` and `TabNote` stem lifecycles are very distinct. So I'm going to make them as consistent as possible.

**Comment by 0xfe** [comment]
Ha, looks like it did fix a bug. :-)

**Comment by Silverwolf90** [comment]
Ok, things should be back to normal-ish. But almost everything will have minor changes.

The stem lifecycle between stavenotes and tabnotes is now more consistent.

Started removed a lot of the shifting that compensates for the `Stem.WIDTH`, but couldn't remove it from everywhere.

After fixing a lot of stem logic, stem connections to noteheads got screwed. The solution seemed to be replacing the hard-coded notehead widths the *real* bounding box width.  For now I just replaced the property with a getter, but this is a temporary hack. 

Ultimately, `tables.js` should not be responsible for anything related to glyph metrics. It should only contain data concerning music theory and music engraving semantics . Everywhere in the codebase that accesses this `head_width` should switch to getting the notehead width from a locally instantiated `Glyph`.

**Comment by Silverwolf90** [comment]
Also something is definitely wrong with Travis. We should be seeing tons of diffs.

**Comment by 0xfe** [comment]
This sounds great. I'm still traveling so will try to get to this as soon
as I get a moment.

**Comment by 0xfe** [comment]
Okay, finally able to look at this. Lots of differences in tests, but all sane. The only visual nit is displaced note heads. It seems like this change has shifted them about a pixel away from the stem, making them look slightly detached when you zoom in. Want to take a quick look?

File: tests/stavenote_tests.js [hunk 1]
```diff
@@ -373,7 +373,7 @@ VF.Test.StaveNote = (function() {
       tickContext.setX(10);
       tickContext.setPadding(0);
 
-      equal(tickContext.getWidth(), 16);
+      equal(tickContext.getWidth(), 15.5988);
```
Comment (0xfe) [inline tests/stavenote_tests.js]: This _might_ cause platform-specific errors due to floating point variations. Does qunit have an `almostEqual`? (or perhaps define `amostEqual(value, expectedValue, errorMargin)`).

**Comment by Silverwolf90** [comment]
Ok this is good to go. Added the `almostEqual` assertion and shifted the displaced noteheads so that they overlap the stem.

**Comment by 0xfe** [comment]
All good. Great change!

**Pull request merged**